libherokubuildpack/src/buildpack_output/mod.rs

```diff
@@ -0,0 +1,500 @@
+
+mod ansi_escape;
+mod duration_format;
+pub mod style;
+mod util;
+
+/// See the module docs for example usage.
+#[allow(clippy::module_name_repetitions)]
+#[derive(Debug)]
+pub struct BuildpackOutput<T> {
+    pub(crate) started: Option<Instant>,
+    pub(crate) state: T,
+}
+
+/// Various states for [`BuildpackOutput`] to contain.
+///
+/// The [`BuildpackOutput`] struct acts as an output state machine. These structs
+/// are meant to represent those states.
+#[doc(hidden)]
+pub mod state {
+    use crate::buildpack_output::util::ParagraphInspectWrite;
+    use crate::write::MappedWrite;
+    use std::time::Instant;
+
+    #[derive(Debug)]
+    pub struct NotStarted<W> {
+        pub(crate) write: ParagraphInspectWrite<W>,
+    }
+
+    #[derive(Debug)]
+    pub struct Started<W> {
+        pub(crate) write: ParagraphInspectWrite<W>,
+    }
+
+    #[derive(Debug)]
+    pub struct Section<W> {
+        pub(crate) write: ParagraphInspectWrite<W>,
+    }
+
+    pub struct Stream<W: std::io::Write> {
+        pub(crate) started: Instant,
+        pub(crate) write: MappedWrite<ParagraphInspectWrite<W>>,
+    }
+}
+
+#[doc(hidden)]
+trait AnnounceSupportedState {
+    type Inner: Write;
+
+    fn write_mut(&mut self) -> &mut ParagraphInspectWrite<Self::Inner>;
+}
+
+impl<W> AnnounceSupportedState for state::Section<W>
+where
+    W: Write,
+{
+    type Inner = W;
+
+    fn write_mut(&mut self) -> &mut ParagraphInspectWrite<Self::Inner> {
+        &mut self.write
+    }
+}
+
+impl<W> AnnounceSupportedState for state::Started<W>
+where
+    W: Write,
+{
+    type Inner = W;
+
+    fn write_mut(&mut self) -> &mut ParagraphInspectWrite<Self::Inner> {
+        &mut self.write
+    }
+}
+
+#[allow(private_bounds)]
+impl<S> BuildpackOutput<S>
+where
+    S: AnnounceSupportedState,
+{
+    #[must_use]
+    pub fn warning(mut self, s: impl AsRef<str>) -> BuildpackOutput<S> {
+        self.write_paragraph(YELLOW, s);
+        self
+    }
+
+    #[must_use]
+    pub fn important(mut self, s: impl AsRef<str>) -> BuildpackOutput<S> {
+        self.write_paragraph(CYAN, s);
+        self
+    }
+
+    pub fn error(mut self, s: impl AsRef<str>) {
+        self.write_paragraph(RED, s);
+    }
+
+    fn write_paragraph(&mut self, color: &str, s: impl AsRef<str>) {
+        let io = self.state.write_mut();
+
+        if !io.was_paragraph {
+            writeln_now(io, "");
+        }
+
+        writeln_now(
+            io,
+            ansi_escape::inject_default_ansi_escape(
+                color,
+                prefix_lines(s.as_ref(), |_, line| {
+                    if line.chars().all(char::is_whitespace) {
+                        String::from("!")
+                    } else {
+                        String::from("! ")
+                    }
+                }),
+            ),
+        );
+        writeln_now(io, "");
+    }
+}
+
+impl<W> BuildpackOutput<state::NotStarted<W>>
+where
+    W: Write,
+{
+    pub fn new(io: W) -> Self {
```

Comment:
Could you add rustdocs for all of these public APIs?

Currently the generated docs are completely empty, eg:

<img width="794" alt="Screenshot 2024-02-02 at 15 47 28" src="https://github.com/heroku/libcnb.rs/assets/501702/fc6554f7-4f8e-4e5c-9ce9-33290e1406ba">

Reply:
The prior docs largely referenced the print_style_guide.rs, which also existed as tests like struct visibility. I'll give everything a pass, but might defer some examples to bringing the style guide back in a future PR.

Reply:
I've updated the docs, please take a look.